proto/storage/cluster.proto

```diff
@@ -94,6 +95,56 @@ message AdmissionControllerConfig {
   bool enforce_on_updates = 5;
 }
 
+message ProcessConfig {
+  bool enabled = 1;
+}
+
+message NetworkConnectionConfig {
+  bool enabled = 1;
+  bool aggregate_external = 2;
+}
+
+message NetworkEndpointConfig {
+  bool enabled = 1;
+  bool include_listening_endpoint_processes = 2;
+}
+
+message CollectorFeature {
+  oneof feature {
+    ProcessConfig processes = 1;
+    NetworkConnectionConfig network_connections = 2;
+    NetworkEndpointConfig network_endpoints = 3;
+  }
+}
+
+message CollectorNamespaceConfig {
+  message NamespaceRule {
+    string namespace = 1;
+    storage.MatchType match_type = 2;
+  }
+
+  CollectorFeature feature = 1;
+  repeated NamespaceRule namespace_selection = 2;
+}
+
+// CollectorConfig controls which type of data is reported by collector
+// and processing of that data by collector. For now this configuration
+// controls if processes, network connections, and network endpoints are
+// reported. It also controls if process information is reported with
+// network endpoints and if external IPs are aggregated. External IP
+// aggregation is controlled at the namespace level and in the future
+// other features may also be controlled at the namespace level.
+message CollectorConfig {
+  // The following three are applied at the cluster level.
+  ProcessConfig process_config = 1;
+  NetworkConnectionConfig network_connection_config = 2;
+  NetworkEndpointConfig network_endpoint_config = 3;
+  // The namespace scope config is an array since it controls multiple
+  // features and for each feature there can be different configs for
+  // different sets of namespaces.
+  repeated CollectorNamespaceConfig namespace_scope_config = 4;
```

Comment:
@JoukoVirtanen I am confused by this comment: https://github.com/stackrox/stackrox/pull/12456#issuecomment-2359878622

I thought all features are controlled on the namespace level and not only `NetworkConnectionConfig`. I feel like these changes are too complex for what it achieves and strongly recommend to go a simpler approach. Please note I've read a design doc, reviewed your PR and we had personal conversation and I am still not confident I am understanding it (or I am bad at listening :p).

My understanding is that:

1. A feature for collector can be enabled or disabled
2. Certain features can be configured to only run in specific namespaces

Based on this I've tried to simplified the protos: https://github.com/stackrox/stackrox/pull/12769/files

This has to me has the following advantages:

1. It is clear which configurations exist in `CollectorConfig` with a single look at the `CollectorConfig` message.
2. By looking at a `CollectorConfig` it is clear which config is supported to be configured for namespaces

Reply:
1. I think there is some nuance to this. It depends upon what is considered a feature. Network connection collection and external IP aggregation are both grouped into `NetworkConnectionConfig`. They are different, but related. Is network connections one feature or two? While for now all of the fields are booleans, non-boolean fields can be added in the future.

2. Is correct.

Your simplified approach will not work. With your protobuf definitions you will only be able to provide `NetworkConnectionConfig` only once. This means that you cannot specify different behaviors for different namespaces. What is the `NetworkConnectionConfig` for a namespace not specified by `namespace_regexes`? We cannot assume that is is just the opposite of what is provided in the only `NetworkConnectionConfig`, as there are multiple fields there and there is a possibility of adding non-boolean fields in the future. What if I want to disable the collection of network connections all together for namespace1, enable it for namespace2 and aggregate external IPs, and enable it for namespace3 and not aggregate external IPs?

Reply:
@JoukoVirtanen Would `repeated NetworkConnectionConfig` solve this problem of being able to only define one?

> What is the NetworkConnectionConfig for a namespace not specified by namespace_regexes? We cannot assume that is is just the opposite of what is provided in the only NetworkConnectionConfig, as there are multiple fields there and there is a possibility of adding non-boolean fields in the future. 

Imho these would be reasonable defaults which collector sets, like I think collector does today. Is this related to the simplified configuration?

> What if I want to disable the collection of network connections all together for namespace1, enable it for namespace2 and aggregate external IPs, and enable it for namespace3 and not aggregate external IPs?

Adding repeated, as `repeated NetworkConnectionConfig network_connection_config = 2;` , would work? :thinking: 

```
message NetworkConnectionConfig {
  bool enabled = 1;
  bool aggregate_external = 2;
  repeated string namespace_regexes = 3;
}

message CollectorConfig {
  // The following three are applied at the cluster level.
  ProcessConfig process_config = 1;
  repeated NetworkConnectionConfig network_connection_config = 2;
  NetworkEndpointConfig network_endpoint_config = 3;
}
```

Could you add a YAML / JSON configuration example of your approach how a user configures it? This would definitely help understand the configuration.

Reply:
I had a meeting with @robbycochran and Maria Simon Marcos. We decided that for 4.6 the only thing that would go into collector runtime configuration would be `aggregate_external` and that it would be only configurable at the cluster level. I still want my proposed protobuf definitions to go into 4.7 and we can have a meeting about that if you like.